jax/_src/lax/control_flow.py

```diff
@@ -383,43 +383,76 @@ def _while_loop_batching_rule(args, dims, axis_name, main_type,
+  cconst_dims, bconst_dims, init_dims = split_list(dims, [cond_nconsts, body_nconsts])
 
+  carry_bat = init_bat
   # Fixpoint computation of which carry are batched: either
   # batched from init, or the carry out is batched. Each iteration promotes
-  # at least one carry to batched. We need at most len(carry) iterations,
-  # but we need one last iteration to prepare the jaxpr based on the final
-  # carry_bat.
-  carry_bat = init_bat
+  # at least one carry to batched. We need at most len(carry) iterations to
+  # reach a fixpoint.
   for _ in range(1 + len(carry_bat)):
-    batched = bconst_bat + carry_bat
-    body_jaxpr_batched, carry_bat_out = batching.batch_jaxpr(
-        body_jaxpr, size, batched, instantiate=carry_bat,
+    _, carry_bat_out = batching.batch_jaxpr(
+        body_jaxpr, size, bconst_bat + carry_bat, instantiate=False,
         axis_name=axis_name, main_type=main_type)
-    cond_jaxpr_batched, (pred_bat,) = batching.batch_jaxpr(
-        cond_jaxpr, size, cconst_bat + carry_bat,
-        instantiate=bool(cond_jaxpr.out_avals[0].shape),
-        axis_name=axis_name, main_type=main_type)
-    carry_bat_out = _map(partial(operator.or_, pred_bat), carry_bat_out)
-    if carry_bat_out == carry_bat:
+    if carry_bat == carry_bat_out:
       break
-    else:
-      carry_bat = _map(operator.or_, carry_bat, carry_bat_out)
+    carry_bat = safe_map(operator.or_, carry_bat, carry_bat_out)
   else:
     assert False, "Fixpoint not reached"
 
-  consts, init = split_list(args, [cond_nconsts + body_nconsts])
-  const_dims, init_dims = split_list(dims, [cond_nconsts + body_nconsts])
-  new_consts = [batching.moveaxis(x, d, 0) if d is not batching.not_mapped and d != 0
-                else x for x, d in zip(consts, const_dims)]
-  new_init = [batching.broadcast(x, size, 0) if now_bat and not was_bat
-              else batching.moveaxis(x, d, 0) if now_bat and d != 0 else x
-              for x, d, was_bat, now_bat in zip(init, init_dims, init_bat, carry_bat)]
+  # Knowing how the carry is batched now, we can determine if the predicate is
+  # batched.
+  _, (pred_bat,) = batching.batch_jaxpr(
+      cond_jaxpr, size, cconst_bat + carry_bat, instantiate=False,
+      axis_name=axis_name, main_type=main_type)
+
+  if pred_bat:
+    # If the predicate is batched, we have to batch *all* of the carry
+    # regardless of if the body needs it.
+    carry_bat = [True] * len(carry_bat)
+    carry_dims = [0] * len(carry_bat)
+    body_jaxpr_batched, _ = batching.batch_jaxpr_axes(
+        body_jaxpr, size, bconst_dims + carry_dims,
+        carry_dims, axis_name=axis_name, main_type=main_type)
+    cond_jaxpr_batched, _ = batching.batch_jaxpr_axes(
+        cond_jaxpr, size, cconst_dims + carry_dims, [0],
+        axis_name=axis_name, main_type=main_type)
+  else:
+    # If the predicate is not batched, we need to make sure that the in axes
+    # match the out axes in the batched `body_jaxpr`. Otherwise the predicate's
```

Comment:
I'm not sure if I follow this comment. What are in axes and out axes here? Can you please reword this?

Reply:
Done.